src/algorithms/split_and_merge_algorithm.h

```diff
@@ -0,0 +1,149 @@
+
+  void compute_restricted_gs_data_idx(const unsigned int first_random_idx,
+                                      const unsigned int second_random_idx);
+
+  void compute_restricted_gs_unique_values(
+      const unsigned int first_random_idx,
+      const unsigned int second_random_idx);
+
+  /* Auxiliary function that computes the log of the ratio of the likelihoods
+   * and of the priors between the proposal and the original clustering.
+   * The two ratios are returned via the reference parameters.
+   * The parameter "split" represents if the function is called in the "split"
+   * case or in the "merge" case.
+   */
+  void compute_log_ratio_like_and_prior(const unsigned int first_random_idx,
+                                        const unsigned int second_random_idx,
+                                        bool split,
+                                        double &log_ratio_prior_prob,
+                                        double &log_ratio_likelihoods);
+
+  void split_or_merge(const unsigned int first_random_idx,
+                      const unsigned int second_random_idx);
+
+  /* Updates unique_values and allocations when a split or a merge proposal
+   * is accepted.
+   */
+  void proposal_update_allocations(const unsigned int first_random_idx,
+                                   const unsigned int second_random_idx,
+                                   const bool split);
+
+  void full_gibbs_sampling();
+
+  /* After picking at random two points, the Split and Merge algorithm
+   * takes all the points that are in the same clusters as the first two and
+   * assigns to each of them, at random, one of two temporary clusters. Then,
+   * it updates these assignments by doing a restricted Gibbs sampling,
+   * restricted in the sense that it considers only the points cited above
+   * and only the two temporary clusters.
+   * The result of this operation is used to compute the acceptance
+   * probability for the split or merge proposal.
+   * For more information, check the article on Split and Merge by
+   * Jain and Neal (2004).
+   *
+   * This function computes one iteration of the restricted Gibbs sampling.
+   *
+   * If return_log_res_prod is true, the function returns the log of the
+   * probability of the transition that restricted_gs_unique_values has done
+   * in the function.
+   * If return_log_res_prod is false, ignore the return value.
+   *
+   * If step_to_original_clust is true, the function moves all the points
+   * in restricted_gs_data_idx to their original clustering configuration.
+   * This step is used in the merge case to compute the transition probability
+   * from the restricted gibbs sampling state to the original split
+   * configuration.
+   */
+  double restricted_gibbs_sampling(const double first_random_idx,
+                                   bool return_log_res_prod = false,
+                                   bool step_to_original_clust = false);
```

Comment:
I'd rather have the big explanation on how the algorithm works at the beginning of the file. Method docstrings should only contain implementation details or a brief summary on what the method does